core/arch/arm/kernel/elf_load.c

```diff
@@ -333,21 +542,52 @@ TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
 	 * Copy the segments
 	 */
 	memcpy(dst, state->ta_head, state->ta_head_size);
-	res = copy_to(state, dst, state->vasize,
-		      phdr[0].p_vaddr + state->ta_head_size,
-		      phdr[0].p_offset + state->ta_head_size,
-		      phdr[0].p_filesz - state->ta_head_size);
-	if (res != TEE_SUCCESS)
-		return res;
+	if (state->is_32bit) {
+		Elf32_Ehdr *ehdr = state->ehdr;
+		Elf32_Phdr *phdr = state->phdr;
+
+		res = copy_to(state, dst, state->vasize,
+			      phdr[0].p_vaddr + state->ta_head_size,
+			      phdr[0].p_offset + state->ta_head_size,
+			      phdr[0].p_filesz - state->ta_head_size);
+		if (res != TEE_SUCCESS)
+			return res;
 
-	for (n = 1; n < ehdr->e_phnum; n++) {
-		if (phdr[n].p_type != PT_LOAD)
-			continue;
+		for (n = 1; n < ehdr->e_phnum; n++) {
+			if (phdr[n].p_type != PT_LOAD)
+				continue;
 
-		res = copy_to(state, dst, state->vasize, phdr[n].p_vaddr,
-			      phdr[n].p_offset, phdr[n].p_filesz);
+			res = copy_to(state, dst, state->vasize,
+				      phdr[n].p_vaddr, phdr[n].p_offset,
+				      phdr[n].p_filesz);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+		shoff = ehdr->e_shoff;
+		shsize = ehdr->e_shnum * sizeof(Elf32_Shdr);
+	} else {
+		Elf64_Ehdr *ehdr = state->ehdr;
```

Comment:
I see your point, but putting this in a macro would make the code harder to read in my opinion.
I'll make an attempt with an access function to read out content of a Phdr based on `is_32bit`.